pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -1060,7 +1060,36 @@ func isKernelPid(pid int) bool {
 	return err != nil && os.IsNotExist(err)
 }
 
+// GetCapacity returns node capacity data for "cpu", "memory", "ephemeral-storage", and "huge-pages*"
+// At present this method is only invoked when introspecting ephemeral storage
 func (cm *containerManagerImpl) GetCapacity() v1.ResourceList {
+	// We store allocatable ephemeral-storage in the capacity property once we Start() the container manager
+	if _, ok := cm.capacity[v1.ResourceEphemeralStorage]; !ok {
+		// If we haven't yet stored the capacity for ephemeral-storage, we can fetch it directly from cAdvisor,
+		// with some retry tolerance in case it hasn't yet started
+		ret := v1.ResourceList{}
+		err := wait.PollImmediate(1*time.Second, 30*time.Second, func() (bool, error) {
```

Comment:
does it mean that we can have here an additional delay during the node status update? Can we just do the same thing without introducing `wait.PollImmediate`?

Reply:
Yes, it means that *if* the initial call to cAdvisor returns an error (or if `cm.cadvisorInterface` is nil, but I don't think that's actually possible in practice — I put it there to be more defensive).

My thinking is that there are edge cases where either cAdvisor isn't yet ready to report a sane response, and for 99% of those edge cases an additional retry or two will produce the desired outcome.

Are you concerned that there are going to be terminal edge cases where we delay node registrations by 30 seconds?

Reply:
If I'm reading this right:

- This could block every time we call `GetCapacity`. Thus, as written this could delay node startup time up to 30s, and potentially block in other places where we didn't expect it to. I'm not sure this should block at all; how many attempts to read cadvisor do we want to make, realistically?
- In case of an error, we return an entirely empty resource list. But the bug is limited to ephemeral storage capacity being set wrong. Hence, I think this change should be limited to updating ephemeral storage capacity, and not return an empty resource list if that can't be found.
- A unit test could have potentially caught that accidental wipe of the rest of capacity info on error, so I suggest we add one as part of this PR.

/lgtm cancel

Reply:
Happy to lean into UT here.

For clarity, this func is *only* invoked here:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/nodestatus/setters.go#L332

Returning an empty (important: not nil) fulfills the existing client usage expectations here:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/nodestatus/setters.go#L334

I can't answer the question about how many times to call cAdvisor. I can do some tests w/ the async + retry removed, just calling it once as a best effort and see if I can still detect the race condition outcome (node registered w/ zero ephemeral storage, thus blocking forward progress for static pod admission on kubeadm + etcd +2nd node scenario).

How does that sound?

(And yes, when this is all done, make sure I include a sane UT that captures the intent :) )

Reply:
Interesting. This is feature gated. That feature is declared as part of https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1029-ephemeral-storage-quotas/README.md which hasn't been updated since 1.16 when it went beta.

This to me seems like an issue with the current KEP implementation.

@kubernetes/sig-storage-misc @msau42 can you PTAL?

Reply:
@ehashman thanks for raising it. We plan to evaluate this feature and make it GA next.